concrete-core-bench/src/benchmark.rs

```diff
@@ -0,0 +1,144 @@
+//! timing loop, and stored in a buffer. The same goes for the post-execution contexts which are all
+//! stored in a buffer and collected out of the timing loop.
+//!
+//! For this reason, it must be possible to store the whole batch of pre/post-execution contexts on
+//! the benchmarked device. Depending on the available memory, this could be problematic. To
+//! accomodate this situation, the input parameter `batch_size` can be used to provide a different
+//! batch size.
+use concrete_core::prelude::AbstractEngine;
+use concrete_core_fixture::fixture::Fixture;
+use concrete_core_fixture::generation::{IntegerPrecision, Maker};
+use criterion::measurement::WallTime;
+use criterion::{BenchmarkGroup, BenchmarkId, Criterion};
+use std::cmp::min;
+use std::ops::Add;
+use std::time::{Duration, Instant};
+
+/// Default value for the batch size.
+pub const DEFAULT_BATCH_SIZE: u64 = 1000;
+
+/// An extension to the `Fixture` trait for benchmarking with criterion.
+pub trait BenchmarkFixture<Precision: IntegerPrecision, Engine: AbstractEngine, RelatedEntities>:
+    Fixture<Precision, Engine, RelatedEntities> + Sized
+{
+    /// Benchmarks all the parameters defined by the fixture.
+    fn bench_all_parameters(
+        maker: &mut Maker,
+        engine: &mut Engine,
+        criterion: &mut Criterion,
+        batch_size: Option<u64>,
+    ) {
+        let mut group = criterion.benchmark_group(format!(
+            "{}<{}, {}, {}>",
+            type_name::<Self>(),
+            type_name::<Precision>(),
+            type_name::<Engine>(),
+            type_name::<RelatedEntities>()
+        ));
+        for params in Self::generate_parameters_iterator() {
+            Self::bench(maker, engine, params, &mut group, batch_size);
+        }
+        group.finish();
+    }
+
+    /// Benchmark one parameter in the fixture benchmark group.
+    fn bench(
+        maker: &mut Maker,
+        engine: &mut Engine,
+        parameters: Self::Parameters,
+        fixture_group: &mut BenchmarkGroup<WallTime>,
+        batch_size: Option<u64>,
+    ) {
+        let batch_size = batch_size.unwrap_or(DEFAULT_BATCH_SIZE);
+
+        // We generate the prototypes once (used for all repetitions of the benchmark).
+        let repetition_proto = Self::generate_random_repetition_prototypes(&parameters, maker);
+        let sample_proto =
+            Self::generate_random_sample_prototypes(&parameters, maker, &repetition_proto);
+
+        // The benchmark itself does not use the standard iterators of `criterion`, because the
+        // pre-context is passed by value, and the post-context must be gathered to be
+        // properly disposed afterward.
+        fixture_group.bench_with_input(
+            BenchmarkId::from_parameter(format!("{:?}", parameters)),
+            &(repetition_proto, sample_proto),
+            |b, (repetition_proto, sample_proto)| {
+                b.iter_custom(|iterations_to_go| {
```

Comment:
How's the value of iterations_to_go set? Is it decided based on criterion parameters?

Reply:
Yes, it is chosen by criterion. Probably based on command-line arguments.